### ISSUE TITLE
Make glibc static for Linux builds

### ISSUE THREAD

**Issue by ReinerBforartists** [Bforartists/Bforartists#2284]
I compile at Ubuntu. And i always keep the environment up to date. Latest Ubuntu version is 20.10

We have currently the problem that Ubuntu 20.10 has a newer glibc version than many other Linux distributions. And so our build just runs at linux distributions that has similar or higher glibc version. And errors out at linux distributions that has a lower glibc version installed. Even Ubuntu 20.04 is not to convince to start Bforartists.

We might drag this problem with us with every new Ubuntu version. And staying at older Ubuntu versions is no option really.

So the idea was to include glibc into the Bforartists binary as a static library. 

Another idea is to bypass this problem by creating a snap or a appimage or flatpak package. They come with static libraries by design. I always failed so far though to create such a package.

**Comment by iyadahmed** [comment]
Wouldn't running an environment with a lower glibc be easier?
Like a Virtual Machine or a container

**Comment by ReinerBforartists** [comment]
Too late, i have already udpated to 20.10 :)
It takes some days to set up yet another Ubuntu version. And as told, we drag the problem with us with ever new Ubuntu version then. So the better option is to make it compatible instead :)

**Comment by ReinerBforartists** [comment]
Not sure if related. But i got some warning in cmake when setting up with static. Blender devs have updated the libs lately ...

![cmakemecker](https://user-images.githubusercontent.com/13919702/110784933-930a4480-826a-11eb-85fc-d333cbd53fc5.jpg)

**Comment by ReinerBforartists** [comment]
As a note, the current Ubuntu  developer version is compiled at Ubuntu 20.04 again. I found an old intact vm ...

**Comment by ReinerBforartists** [comment]
Unassigning for now.

**Comment by seanr** [comment]
There is code in the repo for bul environments and buildbots. Shouldn't we ultimate look into implementing that? I'm not sure what the infrastructure requirements are for that, but in an ideal world, we'd be doing this the same way upstream Blender does.

**Comment by iyadahmed** [comment]
I searched a bit and it seems statically linking glibc is not a good idea, we should just build with a system with same glibc version as official![image](https://user-images.githubusercontent.com/25552173/113164217-31be0b80-9241-11eb-858c-6b96c4e277c7.jpeg)


https://stackoverflow.com/questions/57476533/why-is-statically-linking-glibc-discouraged

https://wiki.blender.org/wiki/Building_Blender/Linux/Generic_Distro

**Comment by naetherm** [comment]
> There is code in the repo for build environments and buildbots. Shouldn't we ultimate look into implementing that? I'm not sure what the infrastructure requirements are for that, but in an ideal world, we'd be doing this the same way upstream Blender does.

I am actually working on that issue. The official Blender binaries are, as @iyadahmed quoted, built on a RHEL7 version. For that to work I am building a CentOS 7 docker image, producing all relevant cmake packages.

**Comment by iyadahmed** [comment]
> > There is code in the repo for build environments and buildbots. Shouldn't we ultimate look into implementing that? I'm not sure what the infrastructure requirements are for that, but in an ideal world, we'd be doing this the same way upstream Blender does.
> 
> 
> 
> I am actually working on that issue. The official Blender binaries are, as @iyadahmed quoted, built on a RHEL7 version. For that to work I am building a CentOS 7 docker image, producing all relevant cmake packages.

Maybe this old docker can help too, or find a newer one, surely someone might have made one

https://hub.docker.com/r/mattiasohlsson/centos-blender-2.8-builder

**Comment by iyadahmed** [comment]
As a note I recently did a build on Debian, documented steps in the process, I also found this interesting project

https://github.com/wheybags/glibc_version_header

**Comment by ReinerBforartists** [comment]
Phew, i am lost with this one. But definitely an interesting find :)

**Comment by iyadahmed** [comment]
> Phew, i am lost with this one. But definitely an interesting find :)

Indeed, but again I think we stick to original plan of compiling on same system as Blender

**Comment by ReinerBforartists** [comment]
Yes, mnaether has created a docker now, based on CentOS. We are currently at testing it :)

**Comment by iyadahmed** [comment]
Awesome

**Comment by ReinerBforartists** [comment]
The tar version is now up in the developer build download section https://www.bforartists.de/download/

https://www.bforartists.de/data/dev/centos-devbuild-08-04-2021.tar.xz

The deb has still a little problem at the moment. It does not add a starter icon. Means you have to navigate to the folder and start it from there, or you need to start it from the console. But when everything else fails then i can do the deb packing at Ubuntu. This way already works with a starter icon. But that's future music. The developer builds are tars. Just he final version comes as a deb.

**Comment by iyadahmed** [comment]
Useful 

https://developer.blender.org/P1013

**Comment by ReinerBforartists** [comment]
Thanks for the link :)

**Comment by naetherm** [comment]
> Useful
> 
> https://developer.blender.org/P1013

Completely out of date. The devtoolset 6 cannot be used for building blender/bforartists as it requires at least version 9 (if I remember it correctly).
The provided script does also nothing more than installing python, cmake, and some more required libraries.
My centos docker container is already working and produces all packages (TGZ, TXZ, DEB), I just have to clean up the Dockerfile and write some more documentation around it.

**Comment by iyadahmed** [comment]
> > Useful
> 
> > 
> 
> > https://developer.blender.org/P1013
> 
> 
> 
> Completely out of date. The devtoolset 6 cannot be used for building blender/bforartists as it requires at least version 9 (if I remember it correctly).
> 
> The provided script does also nothing more than installing python, cmake, and some more required libraries.
> 
> My centos docker container is already working and produces all packages (TGZ, TXZ, DEB), I just have to clean up the Dockerfile and write some more documentation around it. 

Ah ok I just got it from Blender devs, didn't notice the problems

**Comment by ReinerBforartists** [comment]
Completely forgotten about this issue. I will close this issue now. Mnaether has a working docker solution. And he will at least compile the release versions with it. We are also back at Ubutu 20.04 for compiling the devbuild. Means this task is fulfilled. Thanks friends :)
